### PR TITLE
Caches from director

### PR DESCRIPTION
**Pull request by jhiemstrawisc**
This PR incorporates some retooling of the osdf-client/stashcp that allows the client to request its list of ordered caches for a given namespace when downloading a file by performing an HTTP request to an OSDF Director instance. To use an instance of the director instead of the default `namespaces.json`, an env var `OSDF_DIRECTOR_URL` is set.

A few notes: 
- There's a smidge more work to do to completely decouple the client from using the `namespaces.json` when the director is set. In particular, the "cache" struct and any functions that expect it as input could use a bit of reworking. This is because the cache struct expects to hold both the non auth/ auth endpoints and a resource name, and the appropriate URL is later selected by determining whether the namespace requires authentication. Since the Director already does this selection and only returns the required URL, the cache struct gets populated with some data that won't be used, because it already knows whether the auth/non auth URL will be selected according to the information in `namespaces.json`
- Currently, when using the director, the info in `cache.Resource` is set to the cache priority as returned by the director, instead of the actual resource name. At the moment, the Director is agnostic of resource names, so if this is something that it should be returning in its headers, that will need to be changed.

### CODE REVIEW

**Comment by matyasselmeci** [comment]
Can you write some unit tests for this?

**Comment by jhiemstrawisc** [comment]
Sure, I'll write some tests and push them soon.

**Comment by jhiemstrawisc** [comment]
@matyasselmeci I see that this failed a lint test, but I'm having trouble figuring out exactly *what* failed the test. Is there a way to access more detailed output from the linter action that's being used?

**Comment by matyasselmeci** [comment]
Looks like the linter embedded the messages right into the Files view on the GitHub web interface. That's kinda neat. There's a complaint in handle_http.go and another in main.go.

**Comment by jhiemstrawisc** [comment]
Nifty! Now that the linter is happy, is there anything else you can see that needs review before merging?

**Comment by matyasselmeci** [comment]
@djw8605 should take a look once he gets back from CHEP.

**Comment by jhiemstrawisc** [comment]
@matyasselmeci @djw8605 just a friendly ping to see where we are with this. Anything else we'd like to see at this point to finish it up?

File: director.go
Comment (djw8605) [inline director.go:46]: I think this is fine, but I see the letters "NSF" in here like this, and I think of governmental agencies.

File: director.go
Comment (djw8605) [inline director.go:47]: I get why you named the variable this, but it doesn't conform to any variable naming scheme.

File: director.go
Comment (djw8605) [inline director.go:123]: At some point, we may want TLS on every transfer, even if they are not authenticated.  But this is probably a fine assumption now, it won't be in the future.  Where is cache.AuthedReq used later?

File: handle_http.go
Comment (djw8605) [inline handle_http.go:231]: I encourage you to think deeply about how you can restructure this large if statement.  I feel both inside the if and else are doing very similar things, just using different data sources for the caches.  I think with 2 if's instead of 1 large one, you can reduce code duplication significantly.

**Review by djw8605 — CHANGES_REQUESTED** [review]
A few changes for code readability.

**Comment by bbockelm** [comment]
@djw8605 / @jhiemstrawisc -- is this ready to go?

**Comment by jhiemstrawisc** [comment]
Looks like there was a tiny merge conflict from the new credential generation stuff, but that should be squared away now. I was waiting for someone to drop a final LGTM before merging. @djw8605, @bbockelm, are we good to go?

File: director_test.go
Comment (djw8605) [inline director_test.go:166]: This function is using 4 spaces rather than a tab.  Fix the indentation, and I think we're good.

**Comment by djw8605** [comment]
Alright, merging.  Test this like crazy!

**Review by djw8605 — APPROVED** [review]

**Pull request merged**
